tooling/lsp/src/requests/inlay_hint.rs

```diff
@@ -470,16 +607,106 @@ fn push_type_variable_parts(
     }
 }
 
+fn get_expression_name(expression: &Expression) -> Option<String> {
+    match &expression.kind {
+        ExpressionKind::Variable(path, _) => Some(path.last_segment().to_string()),
+        ExpressionKind::Prefix(prefix) => get_expression_name(&prefix.rhs),
+        ExpressionKind::MemberAccess(member_access) => Some(member_access.rhs.to_string()),
+        ExpressionKind::Call(call) => get_expression_name(&call.func),
+        ExpressionKind::MethodCall(method_call) => Some(method_call.method_name.to_string()),
+        ExpressionKind::Cast(cast) => get_expression_name(&cast.lhs),
+        ExpressionKind::Parenthesized(expr) => get_expression_name(expr),
+        ExpressionKind::Constructor(..)
+        | ExpressionKind::Infix(..)
+        | ExpressionKind::Index(..)
```

Comment:
Could we also return a name for `Index`? I think perhaps `Constructor` (considered to be a "function"), `Infix`, `Resolved` may also have sensible names here?

Reply:
> Then for index it would be `foo(array[0])` so we'd hide it if the argument name is `array`? I'm not sure the name relationship would be clear.

I'm not sure of the relationship rust analyzer takes but I do get a hint for `foo(array[0])` if the arg name is `array`:
<img width="217" alt="Screenshot 2024-07-18 at 2 48 01 PM" src="https://github.com/user-attachments/assets/633f79f8-787c-4240-9ede-f95988876571">

Reply:
I don't get a hint with the name `value`, but I do for any other name:
<img width="207" alt="Screenshot 2024-07-18 at 2 49 14 PM" src="https://github.com/user-attachments/assets/4aeb152f-dfb3-414b-86f5-2c12c58ae75e">

Reply:
> I don't get a hint with the name value

Right, there's a rule that if a function has only one parameter and the name is "other" or "value", the hint is not shown. I copied it from Rust Analyzer:

https://github.com/rust-lang/rust-analyzer/blob/6b8b8ff4c56118ddee6c531cde06add1aad4a6af/crates/ide/src/inlay_hints/param_name.rs#L199-L200

Rust Analyzer actually hides it for a couple more names, like "rhs" and "pat", which I guess are common in their standard library.

That said, happy to remove this rule (I honestly wasn't sure about it, but maybe if the name is "other" or "value" then that actual name is not that important).

Reply:
A bit more: I'm even thinking that in any function call with just one argument, the parameter name is maybe irrelevant. For example:

```rust
array.push(value); // becomes array.push(elem: value)
```

That said, maybe in some functions the name can be useful, if you aren't sure what's the thing you are supposed to pass. Alternatively we could eventually add a setting to better control how these things work.

Reply:
Thanks for the explanation. I do lean towards then just always displaying the hint for the `Index` case.

Reply:
Do you mean, in `foo(array[0])` it should display `foo(value: array[0])`?

Reply:
> Then for index it would be foo(array[0]) so we'd hide it if the argument name is array

This is the comment I was referencing, apologies that was not very clear. You can keep as hiding the hints in the case of `other` and `value`.

Reply:
> Then for index it would be foo(array[0]) so we'd hide it if the argument name is array

Just popping in but my opinion in this case would be we don't hide the argument name here since we're passing an array element and not `array` itself. In general though I'm usually very conservative in displaying name hints - I usually like them only with literals.